### PR TITLE
Add toggle-able "debug" overlay

### PR DESCRIPTION
**Pull request by Vilsol**
Adds a "debug" overlay that makes it easier to see the flow of resources through your buildings.

WIP Screenshot:

I am open to any suggestions / edits you would like me to make.

* **Junctions:** Blue arrows vertically, Green horizontally
* **Routers:** White arrows in every direction
* **Overflow:** White arrows in primary direction, Red in overflow directions
* **Sort:** TODO

### CODE REVIEW

File: core/src/io/anuke/mindustry/graphics/BlockRenderer.java [hunk 1]
```diff
@@ -241,6 +241,10 @@ public void processBlocks(){
                             }
                         }
                     }
+
+                    if (Core.settings.getBool("debug_overlay")) {
```
Comment (Quezler) [inline core/src/io/anuke/mindustry/graphics/BlockRenderer.java]: `if(Core.settings.getBool("debug_overlay")){`

File: core/src/io/anuke/mindustry/graphics/Layer.java [hunk 2]
```diff
@@ -12,5 +12,7 @@
     /** Power lasers. */
     power,
     /** Extra layer that's always on top.*/
-    lights
+    lights,
+    /** Debug layer with directional arrows and other informative indicators. */
+    debug
```
Comment (Quezler) [inline core/src/io/anuke/mindustry/graphics/Layer.java]: Maybe an optional trailing `,`?

File: core/src/io/anuke/mindustry/input/DesktopInput.java [hunk 3]
```diff
@@ -10,6 +10,7 @@
 import io.anuke.arc.scene.ui.*;
 import io.anuke.arc.scene.ui.layout.*;
 import io.anuke.arc.util.ArcAnnotate.*;
+import io.anuke.arc.util.Log;
```
Comment (Quezler) [inline core/src/io/anuke/mindustry/input/DesktopInput.java]: Not used anymore according to the diff

File: core/src/io/anuke/mindustry/world/blocks/distribution/Conveyor.java [hunk 4]
```diff
@@ -71,6 +73,76 @@ public void load(){
...
+    public void drawDebug(Tile tile){
+        super.drawDebug(tile);
+
+        int selfRotation = tile.rotation() * 90 - 90;
+        boolean selfShadow = false;
+
+        for(Tile proximityTile : tile.entity.proximity()){
+            byte relativeDirection = tile.relativeTo(proximityTile);
+
+            if (relativeDirection == -1) {
```
Comment (Quezler) [inline core/src/io/anuke/mindustry/world/blocks/distribution/Conveyor.java]: `if(relativeDirection == -1){`

File: core/src/io/anuke/mindustry/world/blocks/distribution/Conveyor.java [hunk 5]
```diff
@@ -71,6 +73,76 @@ public void load(){
...
+                        if(proximityTile.rotation() == proximityTile.relativeTo(tile)){
+                            Draw.rect(overlayArrowBodyRegion, tile.drawx(), tile.drawy(), drawRotation);
+                        }
+                    }else{
+                        Draw.rect(overlayArrowBodyRegion, tile.drawx(), tile.drawy(), drawRotation);
+                    }
+                }
+
+                if(proximityTile.block() instanceof Junction){
+                    if (drawRotation == selfRotation) {
```
Comment (Quezler) [inline core/src/io/anuke/mindustry/world/blocks/distribution/Conveyor.java]: `if(drawRotation == selfRotation){`

File: core/src/io/anuke/mindustry/world/blocks/distribution/Conveyor.java [hunk 6]
```diff
@@ -71,6 +73,76 @@ public void load(){
...
+                        selfShadow = true;
+                    } else {
```
Comment (Quezler) [inline core/src/io/anuke/mindustry/world/blocks/distribution/Conveyor.java]: `}else{`

File: core/src/io/anuke/mindustry/world/blocks/distribution/Conveyor.java [hunk 7]
```diff
@@ -71,6 +73,76 @@ public void load(){
...
+
+        Draw.color(1, 1, 1, 0.8f);
+        Draw.rect(overlayArrowBodyRegion, tile.drawx(), tile.drawy(), selfRotation);
+
+        if(!tile.front().block().outputsItems()
+        || (tile.front().block() instanceof Conveyor && tile.front().rotation() == tile.front().relativeTo(tile))){
+            Draw.rect(overlayArrowRegion, tile.drawx(), tile.drawy(), selfRotation);
+        }
+
+        if (selfShadow) {
```
Comment (Quezler) [inline core/src/io/anuke/mindustry/world/blocks/distribution/Conveyor.java]: `if(selfShadow){`

File: core/src/io/anuke/mindustry/world/blocks/distribution/Junction.java [hunk 8]
```diff
@@ -30,6 +34,58 @@ public Junction(String name){
...
+    @Override
+    public void drawDebug(Tile tile){
+        super.drawDebug(tile);
+
+        // Calculate proximity tiles
+        int proximityTiles = 0;
+        int proximityFade = 0;
+        for (Tile proximityTile : tile.entity.proximity()) {
+            byte relativePosition = tile.relativeTo(proximityTile);
+            if (proximityTile.block().outputsItems()) {
```
Comment (Quezler) [inline core/src/io/anuke/mindustry/world/blocks/distribution/Junction.java]: `for(Tile proximityTile : tile.entity.proximity()){`
&
`if(proximityTile.block().outputsItems()){`

File: core/src/io/anuke/mindustry/world/blocks/distribution/Junction.java [hunk 9]
```diff
@@ -30,6 +34,58 @@ public Junction(String name){
...
+            }
+        }
+
+        // Loop over 4 cardinal directions
+        for (int i = 0; i < 4; i++) {
+            if (i % 2 == 0) {
+                Draw.color(0.2f, 0.8f, 0.2f, 0.8f);
+            } else {
+                Draw.color(0.2f, 0.2f, 0.8f, 0.8f);
+            }
```
Comment (Quezler) [inline core/src/io/anuke/mindustry/world/blocks/distribution/Junction.java]:
```
        // Loop over 4 cardinal directions
        for(int i = 0; i < 4; i++){
            if(i % 2 == 0){
                Draw.color(0.2f, 0.8f, 0.2f, 0.8f);
            }else{
                Draw.color(0.2f, 0.2f, 0.8f, 0.8f);
            }
```

File: core/src/io/anuke/mindustry/world/blocks/distribution/Junction.java [hunk 10]
```diff
@@ -30,6 +34,58 @@ public Junction(String name){
...
+
+            // Always draw horizontal and vertical arrow bodies
+            Draw.rect(overlayArrowBodyRegion, tile.drawx(), tile.drawy(), 90 * i - 90);
+
+            if ((proximityTiles & 1 << i) == 0) {
+                Draw.rect(overlayArrowRegion, tile.drawx(), tile.drawy(), 90 * i - 90);
+            }
+
+            if ((proximityFade & 1 << i) > 0) {
```
Comment (Quezler) [inline core/src/io/anuke/mindustry/world/blocks/distribution/Junction.java]: `if((proximityFade & 1 << i) > 0){`

**Comment by Arkanic** [comment]
mindustry's format has no spaces after the end colon and the start of the curly brace
e.g.
```java
public static void main(String[] args){
    System.out.println("does anyone remember console java :c");
}
```

**Comment by CA6** [comment]
Ok this would be nice if a unforeseen behavior is occurring

**Comment by Anuken** [comment]
While this *could* be a useful feature for new players and specific configurations, I don't think the additional bloat is worth it. A separate mod would be a better place for this feature.

**Pull request closed without merging**